helix-core/src/movement.rs

```diff
@@ -1,184 +1,257 @@
+    })
+}
 
-        if categorize(ch) != categorize(next) {
-            begin += 1;
-        }
+// Helper functions for iterators over (usize, char) tuples
+// (necessary to iterate over ropes efficiently while retaining
+// the index).
+pub trait EnumeratedCharHelpers: Sized {
+    //Returns the index at the current [word/punctuation + whitespace] group
+    fn end_of_block(self) -> Option<usize>;
+    fn end_of_word(self) -> Option<usize>;
+    fn position(self) -> Option<usize>;
+    fn last_position(self) -> Option<usize>;
+    fn at_boundary(self) -> bool;
+    fn skip_newlines(self) -> SkipWhile<Self, NewlineCheck>;
+}
 
-        if !skip_over_next(slice, &mut begin, |ch| ch == '\n') {
-            return None;
-        };
+pub type NewlineCheck = for<'r> fn(&'r (usize, char)) -> bool;
+
+impl<I: Clone + Iterator<Item = (usize, char)>> EnumeratedCharHelpers for I {
+    fn end_of_block(self) -> Option<usize> {
+        let after_newline = self.clone().skip_newlines();
+        let mut pairs = after_newline.clone().zip(after_newline.skip(1));
+        pairs
+            .find_map(|((a_pos, a), (_, b))| {
+                ((categorize(a) != categorize(b)) && (is_end_of_line(b) || !b.is_whitespace()))
+                    .then(|| a_pos)
+            })
+            .or_else(|| self.last_position())
+    }
 
-        end = begin;
+    fn end_of_word(self) -> Option<usize> {
+        let after_newline = self.clone().skip_while(|(_, c)| is_end_of_line(*c));
+        let mut pairs = after_newline.clone().zip(after_newline.skip(1));
+        pairs
+            .find_map(|((a_pos, a), (_, b))| {
+                ((categorize(a) != categorize(b)) && (!a.is_whitespace() || is_end_of_line(b)))
+                    .then(|| a_pos)
+            })
+            .or_else(|| self.last_position())
+    }
 
-        skip_over_next(slice, &mut end, char::is_whitespace);
+    fn last_position(self) -> Option<usize> {
+        self.last().map(|(pos, _)| pos)
+    }
 
-        // refetch
-        let ch = slice.char(end);
+    fn position(mut self) -> Option<usize> {
+        self.next().map(|(pos, _)| pos)
+    }
 
-        if is_word(ch) {
-            skip_over_next(slice, &mut end, is_word);
-        } else if is_punctuation(ch) {
-            skip_over_next(slice, &mut end, is_punctuation);
-        }
+    fn at_boundary(mut self) -> bool {
+        matches!(
+            (self.next(), self.next()),
+            (Some((_, a)), Some((_, b))) if categorize(a) != categorize(b)
+        )
     }
 
-    Some(Range::new(begin, end - 1))
+    fn skip_newlines(self) -> SkipWhile<Self, NewlineCheck> {
+        self.skip_while(|(_, c)| is_end_of_line(*c))
+    }
 }
 
 // ---- util ------------
-
-// used for by-word movement
+/// Returns a forward and backwards iterator over (usize, char), where the first element
+/// always corresponds to the absolute index of the character in the slice.
+pub fn enumerated_chars<'a>(
```

Comment:
I see the usage for this function is always only one of the returned output is used, I think we can split this into two functions, one for forward and one for backward.